### ISSUE TITLE
Composition Stability Issues

### ISSUE THREAD

**Issue by kferrone** [crossplane/crossplane#2867]
### What happened?

I have a repository with a number of compositions creating AWS resources. Some compositions have become increasingly advanced. My overall main issue with the compositions is the absolute lack of any logs, errors, or events. When issues arise for me in the compositions, they simply do nothing, say nothing, and get stuck with condition ready false. It's currently impossible to simply read an error and know what went wrong. Sometimes issues really are something I did wrong. Sometimes the issues really are something unexpected happening with Crossplane. Without any errors or logs all I have is the raw yamls to show what is happening.  To get anything working requires quite the trial and error session to widdle down the yaml to something that does work and work your way back up till you see an issue. 

Basically the effect of any issue simply results in some resource in the composition to be blocked. I have one composition that refuses to create a very simple bucket. One won't create one of my XR resources. Another just does nothing no matter what I do. 

I'm hoping my repository is enough to recreate the problem. 

### How can we reproduce it?

I have shared my repository with some of the crossplane team. 

### What environment did it happen in?
Crossplane version: 1.6.1
Provider aws: v0.23.0
EKS: latest

**Comment by muvaf** [comment]
Don't events on composite resoure (XR) help? I think that reports what is rendered or not.

**Comment by kferrone** [comment]
There are no Events on XR resources when there is an issue for me. I would love to get events from them. The status never shows up either. Last time I experienced this was because one single word was misspelled. Absolutely nothing happens, then I correct mistake and like magic everything appears immediately with events and statuses and all. This makes debugging extremely tedious and error prone because all I have to work on is trial and error. 

I've experienced these issues on EKS and my own home lab k3s as well. Wether it's, GCP provider, AWS provider, or even the SQL provider, always the same. I will see plenty of events, errors, and warnings so long as the resources are created first. 

In one instance, I tried to create an XR within another composition, I have yet to discover anything wrong with it. The composition just freezes and creates nothing even though the XR is the very last item to be created.  Once I remove the XR, all the items in the compositions resources get created pretty quickly. Blocked on this one because one of the resources has a value the XR would not be able to get any other way.  

I have battled with these issues for weeks. I don't understand why I am the only one experiencing these issues??? I experience them quite consistently and across environments too. This makes me feel I am doing something wrong but I have yet to determine what that is???

**Comment by muvaf** [comment]
That's interesting. We get such reports when people confuse claim with composite because claims don't get the events but composites do. We [record an event](https://github.com/crossplane/crossplane/blob/a0daaea/internal/controller/apiextensions/composite/reconciler.go#L475) if we couldn't render, which includes dry-run call. Then if the actual creation fails, we [record](https://github.com/crossplane/crossplane/blob/a0daaea/internal/controller/apiextensions/composite/reconciler.go#L512) another event. 

Would you mind sharing a small example that would let us reproduce?

**Comment by kferrone** [comment]
I have shared the repo with @hasheddan and @haarchi and I will share with you too. I've sent an invite. 

My current issue is my composition called cluster which makes an eks cluster. After the cluster is created it is supposed to put the oidc url to the status of my resource like so:

```yaml
    - type: ToCompositeFieldPath
      fromFieldPath: status.atProvider.identity.oidc.issuer
      toFieldPath: status.oidcUrl
``` 

Then the OpenIDConnectProvider will use this URL like so:

```yaml
    - fromFieldPath: status.oidcUrl
      toFieldPath: spec.forProvider.url
```

Once the cluster is created, the composition just freezes and will never become ready and the OIDC url is never added to the status and therefore the OpenIDConnectProvider is never created. However, once I comment out the section in the compositions resources creating the OIDC provider, the cluster will become ready immediately and the OIDC url is put onto the status. Then I simply uncomment the section and voila the OpenIDConnectProvider appears perfectly how it should be and my webhook perfectly set the thumbprint. Obviously there is something wrong if removing resources and adding them back as is forces crossplane to work all of a sudden. 

Here is where it gets strange. I have a Mutating webhook which adds the thumbprint to the OpenIDConnectProvider. This works very well so long as the OpenIDConnectProvider is actually created. The entire time, before, during, and after the eks creation I can see my logs from the webhook saying it is indeed reviewing the OIDC provider but will do nothing because the url is not set yet. 

Well this means something after my webhook is silently denying the admission to the cluster. My guess is Crossplane since it is a crossplane resource.

**Comment by kferrone** [comment]
So it gets deeper. I guess there is an error in a weird way with the OIDC provider. 
I used the logs from my webhook to get the exact resource that is trying to get created. 
I submitted myself and low and behold a real error 

> The OpenIDConnectProvider "pecan" is invalid: spec.forProvider.thumbprintList: Invalid value: 0: spec.forProvider.thumbprintList in body should have at least 1 items

This makes sense though because the URL has not been set and therefore my webhook will skip over it too. The URL can only be set though if the Cluster resource sets the oidc issuer url on the status of the composite resource, ie the two snippets in the comment above. 

This does show off how the errors do not actually turn into events nor are they seen as a status either.

**Comment by muvaf** [comment]
Alright this is more complex to reproduce than I expect. It seems like somehow the creation fails but error is not returned from the client for some reason. If you have Go experience, one thing I can suggest is to scale down Crossplane to 0 pods, run it in debugger and see what's returned in this call for each of the resources at [this line](https://github.com/crossplane/crossplane/blob/release-1.6/internal/controller/apiextensions/composite/reconciler.go#L509).

Another thing to consider is to make such propagations with `policy.fromFieldPath: Required` so that Crossplane doesn't try to create the resource if the patch wasn't successful.

And for other smaller issues around typos and stuff, Upbound just released [VSCode language server](https://marketplace.visualstudio.com/items?itemName=Upboundio.up) for compositions that will greatly help by giving you immediate feedback.

**Comment by kferrone** [comment]
OK so this helped a lot actually: `policy.fromFieldPath: Required`

Now I was able to move a few steps forward. Now I'm stuck trying to get a nodegroup to to actually connect. 
Yeesh, what a battle
